core/state_processor.go

```diff
@@ -170,14 +194,22 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 						return nil, nil, 0, fmt.Errorf("error decoding bytes %x: %w", stIt.Slot(), err)
 					}
 					copy(safeValue[32-len(value):], value)
-					slotnr := rawdb.ReadPreimage(statedb.Database().DiskDB(), stIt.Hash())
 
-					mkv.addStorageSlot(addr, slotnr, safeValue[:])
+					var slotnr [32]byte
+					if _, err := io.ReadFull(fpreimages, slotnr[:]); err != nil {
+						return nil, nil, 0, fmt.Errorf("reading preimage file: %s", err)
+					}
```

Comment:
Remember that our position in the preimage file always mean "the next preimage to be read", so we basically read 32 bytes which is exactly the preimage we're expecting.